### PR TITLE
Bug 1857162: daemon: inject proxy vars into MCD container

### PR DESCRIPTION
**Pull request by yuqi-zhang**
With the introduction of extensions and reworking of OS updates in
https://github.com/openshift/machine-config-operator/pull/1941,
we now run `oc/podman` directly in the MCD to fetch the OS image.
Thus we need to inject the proxy env vars into the container via the
daemonset definition, otherwise image pulls will fail.

### CODE REVIEW

**Comment by yuqi-zhang** [comment]
Tested that the env vars are injected correctly, and that with the env vars proxy should be able to pull the image. Did not do a full test on a failing proxy cluster, will try to set it up if possible but I'm pretty confident this should work

**Comment by yuqi-zhang** [comment]
/hold

did not pass on an upgrade in proxy, investigating

**Comment by yuqi-zhang** [comment]
Ive tested this to work on AWS again, and it properly synced the env var on an upgrade. Will test on Azure before removing the hold. Also note that we should eventually look to either update or remove our lib code but that's not in scope of this PR

**Comment by yuqi-zhang** [comment]
/hold cancel
/retest

Can confirm this successfully passed an upgrade on an azure proxy env

**Review by sinnykumari — APPROVED** [review]
Looks great, thanks Jerry for fixing it!

/lgtm

**Comment by sinnykumari** [comment]
@yuqi-zhang Curious if we have any public documentation somewhere to run OCP cluster in proxy environment, this can help us in future to do proxy related testing.

**Review by cgwalters — APPROVED** [review]
Overall LGTM but does seem like we should take on switching to library-go after branching for 4.6.

File: lib/resourcemerge/core.go [hunk 1]
```diff
@@ -99,6 +99,23 @@ func ensureContainer(modified *bool, existing *corev1.Container, required corev1
...
+				break
+			}
+		}
+		if existingCurr == nil {
+			*modified = true
+			existing.Env = append(existing.Env, corev1.EnvVar{})
+			existingCurr = &existing.Env[len(existing.Env)-1]
+		}
+		ensureEnvVar(modified, existingCurr, required)
+	}
```
[the remark this answers is not in the document]
Comment (yuqi-zhang) [reply lib/resourcemerge/core.go:117]: Hmm yeah I just copied the `Ports` section below but now you mentioned it we don't check for removals. I think that's a problem for all changes applied to the resource. Maybe merging with upstream library-go will fix this

**Comment by yuqi-zhang** [comment]
> Curious if we have any public documentation somewhere to run OCP cluster in proxy environment, this can help us in future to do proxy related testing.

Unfortunately not really. The AWS cluster-bot proxy env is the closest we have but it doesn't use a real proxy today. I talked to Eric Wolinetz and he is looking to get it updated. For this PR specifically I requested some environments from QE.

**Pull request merged**